dao/src/main/java/com/epam/esm/dao/dao/impl/GiftCertificateDaoImpl.java

```diff
@@ -1,41 +1,144 @@
+        return jdbcTemplate.update(DELETE_BY_ID, id) > 0;
     }
 
     @Override
-    public int create(GiftCertificate entity) {
-        throw new UnsupportedOperationException("Not supported yet.");
+    public boolean deleteLinkTagById(long id) {
+        return jdbcTemplate.update(DELETE_LINK_TAG_BY_ID, id) > 0;
     }
 
     @Override
-    public int update(GiftCertificate entity) {
-        throw new UnsupportedOperationException("Not supported yet.");
+    public boolean addLinkTag(long certificateId, long tagId) {
+        PreparedStatementCreator preparedStatementCreator = connection -> {
+            PreparedStatement preparedStatement = connection.prepareStatement(ADD_LINK_ONE_TAG);
+            preparedStatement.setLong(1, certificateId);
+            preparedStatement.setLong(2, tagId);
+            return preparedStatement;
+        };
+        return jdbcTemplate.update(preparedStatementCreator) > 0;
+    }
+
+    @Override
+    public List<GiftCertificate> findEntityByParams(Map<String, String> groupParams) {
+        String sql = FIND_CERTIFICATES_BY_PARAMS +
+                BuilderSql.buildFindAndSortCertificateByParameter(groupParams);
+        return jdbcTemplate.query(sql, certificateExtractor);
+
     }
+
+    @Override
+    public Optional<GiftCertificate> findByName(String certificateName) {
+        List<GiftCertificate> certificateList = jdbcTemplate.query(FIND_BY_NAME, certificateExtractor, certificateName);
+        assert certificateList != null;
+        return certificateList.isEmpty() ? Optional.empty() : Optional.of(certificateList.get(0));
+    }
+
+    @Override
+    public GiftCertificate create(GiftCertificate entity) {
+        KeyHolder key = new GeneratedKeyHolder();
+        PreparedStatementCreator preparedStatementCreator = connection -> {
+            PreparedStatement preparedStatement = connection.prepareStatement(ADD_NEW_CERTIFICATE,
+                    Statement.RETURN_GENERATED_KEYS);
+            preparedStatement.setString(1, entity.getName());
+            preparedStatement.setString(2, entity.getDescription());
+            preparedStatement.setBigDecimal(3, entity.getPrice());
+            preparedStatement.setInt(4, entity.getDuration());
+            preparedStatement.setTimestamp(5, Timestamp.valueOf(LocalDateTime.now()));
+            preparedStatement.setTimestamp(6, Timestamp.valueOf(LocalDateTime.now()));
+            return preparedStatement;
+        };
+        jdbcTemplate.update(preparedStatementCreator, key);
+        Number id = key.getKey();
+        if (id != null) {
+            entity.setId(id.longValue());
+        }
+        return entity;
+    }
+
+    @Override
+    public GiftCertificate update(GiftCertificate entity) {
+        jdbcTemplate.update(UPDATE_CERTIFICATE, entity.getName(), entity.getDescription(),
+                entity.getPrice(), entity.getDuration(), Timestamp.valueOf(LocalDateTime.now()),
```

Comment:
think `create_date` should not be updated, is set once when certificate is being created

Reply:
done